### PR TITLE
Create a DependencySolver to resolve install order

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
This patch fixes the dependency resolution for the installation order based on the user requests on the Holodeck Env File, the previous implementation is broken when kubernetes is not requested

### CODE REVIEW

File: pkg/provisioner/provisioner.go [hunk 1]
```diff
@@ -71,10 +71,8 @@ func New(log *logger.FunLogger, keyPath, userName, hostUrl string) (*Provisioner
 }
 
 func (p *Provisioner) Run(env v1alpha1.Environment) error {
-	graph, err := buildDependencyGraph(env)
-	if err != nil {
-		return fmt.Errorf("failed to build dependency graph: %v", err)
-	}
+	solver := NewDependencySolver(env)
```
Comment (elezar) [inline pkg/provisioner/provisioner.go:74]: Probably `resolver` vs `solver`?

File: pkg/provisioner/provisioner.go [hunk 2]
```diff
@@ -71,10 +71,8 @@ func New(log *logger.FunLogger, keyPath, userName, hostUrl string) (*Provisioner
...
+	solver.ResolveDependencies()
```
Comment (elezar) [inline pkg/provisioner/provisioner.go:75]: Qustion: Why does `a resolver` not return the dependencies?

File: pkg/provisioner/provisioner.go [hunk 3]
```diff
@@ -88,7 +86,7 @@ func (p *Provisioner) Run(env v1alpha1.Environment) error {
 		env.Spec.Kubernetes.K8sEndpointHost = p.HostUrl
 	}
 
-	for _, node := range graph {
+	for _, node := range solver.Dependencies {
```
Comment (elezar) [inline pkg/provisioner/provisioner.go:89]: Why not `dependencies.Resolve()` here?

File: pkg/provisioner/dependency.go [hunk 4]
```diff
@@ -191,3 +86,83 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
...
+	case "kubeadm":
+		d.Dependencies = append(d.Dependencies, functions["kubeadm"])
+	case "kind":
+		d.Dependencies = append(d.Dependencies, functions["kind"])
+	case "microk8s":
+		// reset the graph to only include microk8s
+		d.Dependencies = nil
+		d.Dependencies = append(d.Dependencies, functions["microk8s"])
+	default:
+		// default to kubeadm if KubernetesInstaller is empty
```
Comment (elezar) [inline pkg/provisioner/dependency.go:113]: Question: Why is the default not handled elsewhere? Should `d.env.Spec.Kubernetes.KubernetesInstaller` not be set to a default before calling this function?

File: pkg/provisioner/dependency.go [hunk 5]
```diff
@@ -191,3 +86,83 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
...
+
+func (d *DependencySolver) withKubernetes() {
+	switch d.env.Spec.Kubernetes.KubernetesInstaller {
+	case "kubeadm":
+		d.Dependencies = append(d.Dependencies, functions["kubeadm"])
+	case "kind":
+		d.Dependencies = append(d.Dependencies, functions["kind"])
+	case "microk8s":
+		// reset the graph to only include microk8s
+		d.Dependencies = nil
```
Comment (elezar) [inline pkg/provisioner/dependency.go:110]: Does using `microk8s` not also depend on the container toolkit?

File: pkg/provisioner/dependency.go [hunk 6]
```diff
@@ -191,3 +86,83 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
...
+
+func (d *DependencySolver) withContainerToolkit() {
+	d.Dependencies = append(d.Dependencies, functions["containerToolkit"])
+}
+
+func (d *DependencySolver) withNVDriver() {
+	d.Dependencies = append(d.Dependencies, functions["nvdriver"])
+}
+
+// buildDependencyGraph builds a dependency graph based on the environment
```
Comment (elezar) [inline pkg/provisioner/dependency.go:140]: This really doesn't seem to build a graph, but just a list. Or am I misunderstanding something?

File: pkg/provisioner/dependency.go [hunk 7]
```diff
@@ -191,3 +86,83 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
...
+// and returns a topologically sorted list of provisioning functions
+// to be executed in an opinionated order. We go bottom up in the graph
+// NVDriver -> Container Runtime -> Container Toolkit -> Kubernetes
+// Kubernetes -> Container Runtime -> Container Toolkit -> NVDriver
+// if a dependency is needed and not defined, we set an opinionated default
```
Comment (elezar) [inline pkg/provisioner/dependency.go:145]: What does `bottom-up` mean in this context? Looking at the example it seems as if ONLY NVDriver and Kubernetes are swapped. Is this correct?

File: pkg/provisioner/provisioner.go [hunk 8]
```diff
@@ -110,6 +108,11 @@ func (p *Provisioner) Run(env v1alpha1.Environment) error {
 		p.tpl.Reset()
 	}
 
+	// Close the current ssh connection
+	if err := p.Client.Close(); err != nil {
+		return fmt.Errorf("failed to close ssh client: %v", err)
+	}
+
```
Comment (elezar) [inline pkg/provisioner/provisioner.go:115]: This change seems unrelated?

File: pkg/provisioner/dependency.go (hunk 5 above)
Comment (ArangoGutierrez) [reply pkg/provisioner/dependency.go:110]: no, is a canonical snap pkg. it simply needs a fresh ubuntu system, and it will do everything

File: pkg/provisioner/dependency.go (hunk 6 above)
Comment (ArangoGutierrez) [reply pkg/provisioner/dependency.go:140]: yeah is a list, I was trying to be fancy, but at the end is a list.

File: pkg/provisioner/dependency.go (hunk 4 above)
Comment (ArangoGutierrez) [reply pkg/provisioner/dependency.go:113]: Until this point we have not modified `d.env` , and after this step is not needed.  I think is ok to have it here. unless there's a reason to set the defaults at an earlier stage

File: pkg/provisioner/provisioner.go (hunk 1 above)
Comment (ArangoGutierrez) [reply pkg/provisioner/provisioner.go:74]: addressed

File: pkg/provisioner/provisioner.go (hunk 2 above)
Comment (ArangoGutierrez) [reply pkg/provisioner/provisioner.go:75]: addressed

File: pkg/provisioner/provisioner.go (hunk 3 above)
Comment (ArangoGutierrez) [reply pkg/provisioner/provisioner.go:89]: addressed

File: pkg/provisioner/dependency.go (hunk 6 above)
Comment (ArangoGutierrez) [reply pkg/provisioner/dependency.go]: addressed

File: pkg/provisioner/provisioner.go (hunk 8 above)
Comment (ArangoGutierrez) [reply pkg/provisioner/provisioner.go]: removed

File: pkg/provisioner/dependency.go [hunk 9]
```diff
@@ -191,3 +86,78 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
...
+
+func NewDependencies(env v1alpha1.Environment) *DependencyResolver {
+	return &DependencyResolver{
+		env: env,
+	}
+}
+
+func (d *DependencyResolver) withKubernetes() {
+	switch d.env.Spec.Kubernetes.KubernetesInstaller {
+	case "kubeadm":
```
Comment (tariq1890) [inline pkg/provisioner/dependency.go:104]: Let's move these strings to constant vars

File: pkg/provisioner/dependency.go [hunk 10]
```diff
@@ -191,3 +86,78 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
...
+}
+
+func (d *DependencyResolver) withNVDriver() {
+	d.Dependencies = append(d.Dependencies, functions["nvdriver"])
+}
+
+// Resolve returns the dependency list in the correct order
+func (d *DependencyResolver) Resolve() []ProvisionFunc {
+	// Add NVDriver to the list
+	if d.env.Spec.NVIDIADriver.Install {
```
Comment (tariq1890) [inline pkg/provisioner/dependency.go:143]: I wonder if we can have a function map instead and we could loop through the keys of the map and invoke each function.

The keys of the function map here would be `["nvdriver", "container-runtime", "container-toolkit", "kubernetes"]`

Comment (ArangoGutierrez) [reply pkg/provisioner/dependency.go:143]: we need it to be an ordered list, to install dependencies in order.

Comment (tariq1890) [reply pkg/provisioner/dependency.go:143]: Sure. The keys can be in a slice which guarantees order and then you can invoke the function map as you iterate through the slice
